File: Echo/IntroViewController.m
//
//  IntroViewController.m
//  Echo
//
//  Created by Will Entriken on 11/24/12.
//
//

#import "IntroViewController.h"
#import "Profile.h"
#import "MBProgressHUD.h"
#import "LanguageSelectController.h"
#import "NetworkManager.h"

@interface IntroViewController() <MBProgressHUDDelegate, LanguageSelectControllerDelegate>
@property (strong, nonatomic) MBProgressHUD *hud;
@end

@implementation IntroViewController
@synthesize hud = _hud;
@synthesize delegate = _delegate;

- (IBAction)languageButtonClicked:(UIButton *)sender
{
    if (sender.tag == 1)
        [self saveLanguageWithTag:@"en"];
    if (sender.tag == 2)
        [self saveLanguageWithTag:@"es"];
    if (sender.tag == 3)
        [self saveLanguageWithTag:@"cmn"];
}

- (void)saveLanguageWithTag:(NSString *)tag
{
    Profile *me = [Profile currentUserProfile];
    me.learningLanguageTag = tag;
    self.hud = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    self.hud.mode = MBProgressHUDModeIndeterminate;
    self.hud.labelText = @"Sending...";
    
    [me syncOnlineOnSuccess:^(NSArray *recommendedLessons)
     {
         for (Lesson *lesson in recommendedLessons)
             [self.delegate downloadLessonViewController:self gotStubLesson:lesson];
         [self.hud hide:YES];
         [me syncToDisk];
         [self dismissViewControllerAnimated:YES completion:nil];
     } onFailure:^(NSError *error) {
         [self.hud hide:NO];
         [NetworkManager hudFlashError:error];
     }];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.screenName = @"IntroView";
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    LanguageSelectController *controller = segue.destinationViewController;
    controller.delegate = self;
}

#pragma mark - MBProgressHUDDelegate

- (void)hudWasHidden:(MBProgressHUD *)hud
{
    self.hud = nil;
}

#pragma mark - LanguageSelectControllerDelegate

- (void)languageSelectController:(id)controller didSelectLanguage:(NSString *)tag withNativeName:(NSString *)name
{
    [self saveLanguageWithTag:tag];
    [controller dismissAnimated:YES];
}

@end

File: Echo/IntroViewController.swift
//
//  IntroViewController.swift
//  Echo
//
//  Created by William Entriken on 12/17/15.
//
//

import Foundation
import UIKit
import MBProgressHUD
import Firebase

class IntroViewController: UIViewController, MBProgressHUDDelegate, LanguageSelectControllerDelegate {
    
    weak var delegate: DownloadLessonDelegate?
    weak var hud: MBProgressHUD?
    
    @IBAction func languageButtonClicked(sender: UIButton) {
        if sender.tag == 1 {
            self.saveLanguageWithTag("en")
        }
        if sender.tag == 2 {
            self.saveLanguageWithTag("es")
        }
        if sender.tag == 3 {
            self.saveLanguageWithTag("cmn")
        }
    }
    
    func saveLanguageWithTag(tag: String) {
        let me = Profile.currentUser
        me.learningLanguageTag = tag
        me.syncToDisk()
        let hud = MBProgressHUD.showHUDAddedTo(self.view!, animated: true)
        hud.mode = MBProgressHUDMode.Indeterminate
        hud.labelText = "Sending..."
        self.hud = hud
        me.syncOnlineOnSuccess({
            (recommendedLessons) -> Void in
            for lesson in recommendedLessons {
                self.delegate?.downloadLessonViewController(self, gotStubLesson: lesson)
            }
            hud.hide(true)
            me.syncToDisk()
            self.dismissViewControllerAnimated(true, completion: nil)

            }) { (error) -> Void in
                hud.hide(false)
                MBProgressHUD.flashError(error)
        }
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        FIRAnalytics.logEventWithName("page_view", parameters: ["name": NSStringFromClass(self.dynamicType)])
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        let controller = segue.destinationViewController as! LanguageSelectController
        controller.delegate = self
    }
    
    func hudWasHidden(hud: MBProgressHUD) {
        self.hud = nil
    }
    
    func languageSelectController(controller: AnyObject, didSelectLanguage tag: String, withNativeName name: String) {
        self.saveLanguageWithTag(tag)
        controller.dismissAnimated(true)
    }
}